pkg/solana/chainwriter/lookups.go

```diff
@@ -0,0 +1,354 @@
+	for _, publicKeyMeta := range publicKeys {
+		address, _, err := solana.FindProgramAddress(seeds, publicKeyMeta.PublicKey)
+		if err != nil {
+			return nil, fmt.Errorf("error finding program address: %w", err)
+		}
+		addresses = append(addresses, &solana.AccountMeta{
+			PublicKey:  address,
+			IsSigner:   lookup.IsSigner,
+			IsWritable: lookup.IsWritable,
+		})
+	}
+	return addresses, nil
+}
+
+func (s *SolanaChainWriterService) ResolveLookupTables(ctx context.Context, args any, lookupTables LookupTables) (map[string]map[string][]*solana.AccountMeta, map[solana.PublicKey]solana.PublicKeySlice, error) {
+	derivedTableMap := make(map[string]map[string][]*solana.AccountMeta)
+	staticTableMap := make(map[solana.PublicKey]solana.PublicKeySlice)
+
+	// Read derived lookup tables
+	for _, derivedLookup := range lookupTables.DerivedLookupTables {
+		// Load the lookup table - note: This could be multiple tables if the lookup is a PDALookups that resovles to more
+		// than one address
+		lookupTableMap, _, err := s.LoadTable(ctx, args, derivedLookup, s.reader, derivedTableMap)
+		if err != nil {
+			return nil, nil, fmt.Errorf("error loading derived lookup table: %w", err)
+		}
+
+		// Merge the loaded table map into the result
+		for tableName, innerMap := range lookupTableMap {
+			if derivedTableMap[tableName] == nil {
+				derivedTableMap[tableName] = make(map[string][]*solana.AccountMeta)
+			}
+			for accountKey, metas := range innerMap {
+				derivedTableMap[tableName][accountKey] = metas
+			}
+		}
+	}
+
+	// Read static lookup tables
+	for _, staticTable := range lookupTables.StaticLookupTables {
+		addressses, err := getLookupTableAddresses(ctx, s.reader, staticTable)
+		if err != nil {
+			return nil, nil, fmt.Errorf("error fetching static lookup table address: %w", err)
+		}
+		staticTableMap[staticTable] = addressses
+	}
+
+	return derivedTableMap, staticTableMap, nil
+}
+
+func (s *SolanaChainWriterService) LoadTable(ctx context.Context, args any, rlt DerivedLookupTable, reader client.Reader, derivedTableMap map[string]map[string][]*solana.AccountMeta) (map[string]map[string][]*solana.AccountMeta, []*solana.AccountMeta, error) {
+	// Resolve all addresses specified by the identifier
+	lookupTableAddresses, err := GetAddresses(ctx, args, []Lookup{rlt.Accounts}, nil, reader)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error resolving addresses for lookup table: %w", err)
+	}
+
+	// Nested map in case the lookup table resolves to multiple addresses
+	resultMap := make(map[string]map[string][]*solana.AccountMeta)
+	var lookupTableMetas []*solana.AccountMeta
+
+	// Iterate over each address of the lookup table
+	for _, addressMeta := range lookupTableAddresses {
+		// Read the full list of addresses from the lookup table
+		addresses, err := getLookupTableAddresses(ctx, reader, addressMeta.PublicKey)
+		if err != nil {
+			return nil, nil, fmt.Errorf("error fetching lookup table address: %w", err)
```

Comment:
does it make sense to also log the table address here?

Reply:
Good call, fixed.